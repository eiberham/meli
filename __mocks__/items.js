export const items = [{
    "id": "MLA619208970",
    "title": "Remera Lisa Algodón Jersey Peinado Premium",
    "condition": "new",
    "picture": "http://mla-s1-p.mlstatic.com/876978-MLA31035095700_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 259,
        "decimals": ""
    }
}, {
    "id": "MLA737922842",
    "title": "Remeras Personalizadas Estampada Sublimada Tu Logo",
    "condition": "new",
    "picture": "http://mla-s1-p.mlstatic.com/974505-MLA31102958584_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 199.99,
        "decimals": "99"
    }
}, {
    "id": "MLA752716977",
    "title": "Remeras De Algodon Premium Al Por Mayor",
    "condition": "new",
    "picture": "http://mla-s2-p.mlstatic.com/828346-MLA31115819250_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 260,
        "decimals": ""
    }
}, {
    "id": "MLA670757536",
    "title": "Remeras Entalladas Hombre De Importación Slim Fit!!",
    "condition": "new",
    "picture": "http://mla-s1-p.mlstatic.com/708594-MLA31133445654_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 249.99,
        "decimals": "99"
    }
}];