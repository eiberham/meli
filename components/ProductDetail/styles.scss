.item-detail {
    background: #fff;
    min-height: 80vh;
    display: grid;
    grid-template-columns: 55vw 1fr;
    grid-template-rows: repeat(2, min-content);
    grid-gap: 0;
    grid-template-areas: "image price"
                         "description price";

    &__image {
        grid-area: image;
    }

    &__description {
        grid-area: description;

        p {
            text-align: justify;
        }
    }

    &__price {
        grid-area: price;
    }
}