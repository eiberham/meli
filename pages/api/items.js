export default (req, res) => {
    console.log("el objeto req negro: ", req);
    res.send("Todo bien papu");
}