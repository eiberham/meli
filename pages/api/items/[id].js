import redis from 'redis';
import axios from 'axios';
import _ from 'lodash';
import meli from '../../../services/meli';

export default (req, res) => {
    const client = redis.createClient();
    let id = req.query.id;

    client.get(id, async (err, result) => {
        if (result) {
            res.send(result);
        } else {
            axios.all([
                meli.get(`/items/${id}`),
                meli.get(`/items/${id}/description`)
            ])
                .then(axios.spread(async (first, second) => {
                    let response = {...first.data, ...second.data};

                    let data = {
                        ..._.pick(response, ['id', 'title', 'condition', 'sold_quantity']),
                        picture: _.get(response, 'thumbnail'),
                        free_shipping: _.get(response, 'shipping.free_shipping'),
                        description: _.get(response, 'plain_text'),
                        price: {
                            currency: _.get(response, 'currency_id'),
                            amount: _.get(response, 'price'),
                            decimals: _.toString(_.split(_.get(response, 'price'), '.')[1])
                        }
                    };

                    const {data:{path_from_root}} = await meli.get(`/categories/${response.category_id}`);
                    let categories = _.map(path_from_root, item => {
                        return item.name
                    });

                    let output = {
                        author: { name: 'Abraham', lastname: 'Cedeno'},
                        item: data,
                        categories
                    };

                    client.setex(id, 3600, JSON.stringify(output));

                    res.status(200).json(output);
                }))
                .catch(({response:{data}}) => {
                    if(data.status === 404){
                        return res.status(404).json({
                            "status": 404,
                            "message": data.message
                        });
                    }
                });
        }
    });
}