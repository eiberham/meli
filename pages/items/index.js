import { Fragment } from 'react';
import fetch from 'isomorphic-unfetch';
import Layout from '../../components/Layout';
import Breadcrumb from '../../components/Breadcrumb';
import ProductList from '../../components/ProductList';
import Loading from '../../components/Loading';
import { withRouter } from 'next/router';

const Items = props => {
    const { items, router, categories } = props;

    const onItemSelect = (id) => {
        router.push(`/items/${id}`)
    }

    return (
        <Layout>
            { (items) ? (
                <Fragment>
                    <Breadcrumb links={categories} />
                    <ProductList items={items} onItemSelect={onItemSelect}/>
                </Fragment>
            ) : (
                <Loading size="massive" show />
            )}
        </Layout>
    );
};

Items.getInitialProps = async (context) => {
    const {
        query: { search },
    } = context;
    
    const response = await fetch(`http://localhost:3000/api/items?q=${search}`);
    const data = await response.json();
    return data;
}

export default withRouter(Items);