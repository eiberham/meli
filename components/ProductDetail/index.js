import React from 'react';
import './styles.scss';
import PropTypes from 'prop-types';

const ProductDetail = props => {
    const { item } = props;
    return (
        <React.Fragment>
            <div className="item-detail">
                <div className="item-detail__image">
                    <figure>
                        <img src={item.picture} alt={item.title} />
                    </figure>
                </div>
                <div className="item-detail__description">
                    <h1>Descripci√≥n del producto</h1>
                    <p>{item.description}</p>
                </div>
                <div className="item-detail__price">
                    <span>$ {item.price.amount}</span>
                </div>
            </div>
        </React.Fragment>
    )
}

ProductDetail.propTypes = {
    item: PropTypes.shape({
        title: PropTypes.string,
        id: PropTypes.string,
        condition: PropTypes.string,
        description: PropTypes.string,
        free_shipping: PropTypes.bool,
        picture: PropTypes.string,
        sold_quantity: PropTypes.number,
        price: PropTypes.shape({
            currency: PropTypes.string,
            amount: PropTypes.number,
            decimals: PropTypes.string
        })
    }).isRequired
}

export default ProductDetail;