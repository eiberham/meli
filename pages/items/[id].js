import fetch from 'isomorphic-unfetch';
import Layout from '../../components/Layout';
import Breadcrumb from '../../components/Breadcrumb';
import ProductDetail from '../../components/ProductDetail';

const Item = props => {
    const { item, categories } = props;

    return (
        <Layout>
            <Breadcrumb links={categories} />
            <ProductDetail item={item} />
        </Layout>
    )
}

Item.getInitialProps = async context => {
    const {
        query: { id },
    } = context;
    
    const response = await fetch(`http://localhost:3000/api/items/${id}`);
    const data = await response.json();
    return data;
}

export default Item;