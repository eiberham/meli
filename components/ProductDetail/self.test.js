import React from 'react';
import { items } from '../../__mocks__/items';
import renderer from 'react-test-renderer';
import ProductDetail from './index';

describe('ProductDetail', () => {
    it('matches snapshot', () => {
        const product = items[0];
        const component = renderer.create(<ProductDetail item={product} />)
        .toJSON();
        expect(component).toMatchSnapshot();
    })
})