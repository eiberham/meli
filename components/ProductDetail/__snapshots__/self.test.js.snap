// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail matches snapshot 1`] = `
<div
  className="item-detail"
>
  <div
    className="item-detail__image"
  >
    <figure>
      <img
        alt="Remera Lisa Algodón Jersey Peinado Premium"
        src="http://mla-s1-p.mlstatic.com/876978-MLA31035095700_062019-I.jpg"
      />
    </figure>
  </div>
  <div
    className="item-detail__description"
  >
    <h1>
      Descripción del producto
    </h1>
    <p />
  </div>
  <div
    className="item-detail__summary"
  >
    <span
      className="item-detail__summary-condition"
    >
      Nuevo - undefined vendidos
    </span>
    <span
      className="item-detail__summary-title"
    >
      Remera Lisa Algodón Jersey Peinado Premium
    </span>
    <span
      className="item-detail__summary-amount"
    >
      $ 
      259
    </span>
    <span
      className="item-detail__summary-decimals"
    >
      
    </span>
    <button
      className="ui blue fluid button"
      onClick={[Function]}
    >
      Comprar
    </button>
  </div>
</div>
`;
